tao_compiler/mlir/disc/transforms/disc_dense_to_sparse.cc

```diff
@@ -0,0 +1,423 @@
+  for (int64_t val : transpose_permutation) {
+    transposed_shape.push_back(input_shape[val]);
+  }
+  return RankedTensorType::get(transposed_shape, input_type.getElementType());
+}
+
+Operation* InsertTranspose(
+    mhlo::DotGeneralOp op, Value value,
+    const SmallVectorImpl<int64_t>& transpose_permutation, OpBuilder& b) {
+  auto transpose_permutation_attr =
+      GetI64ElementsAttr(transpose_permutation, &b);
+
+  // Compute the resulting shape.
+  llvm::SmallVector<int64_t, 4> transposed_shape;
+  ShapedType input_type = value.getType().cast<ShapedType>();
+  auto input_shape = input_type.getShape();
+  for (auto val : transpose_permutation) {
+    transposed_shape.push_back(input_shape[val]);
+  }
+  auto transpose_type = GetTransposeOutputType(value, transpose_permutation, b);
+  auto transpose_op = b.create<mhlo::TransposeOp>(
+      op.getLoc(), transpose_type, value, transpose_permutation_attr);
+
+  if (auto attr = op->getAttr(placement_utils::kDiscPlaceAssignment))
+    transpose_op->setAttr(placement_utils::kDiscPlaceAssignment, attr);
+
+  return transpose_op;
+}
+
+// convert:
+//   mhlo.dense_gemm(x, w)
+// to:
+//   mhlo.transpose(x, [1, 0])
+//   mhlo.sparse_gemm(x, w)
+//   mhlo.transpose(y, [1, 0])
+struct ExpandDotGeneralOp : public OpRewritePattern<mhlo::DotGeneralOp> {
+  using OpRewritePattern<mhlo::DotGeneralOp>::OpRewritePattern;
+
+  LogicalResult SparseOp(mhlo::DotGeneralOp op, PatternRewriter& rewriter,
+                         int sparse_weight_pos) const {
+    OpBuilder b(op);
+    Location loc = op.getLoc();
+
+    Value input, kernel;
+    if (sparse_weight_pos == 0) {
+      input = op->getOperand(1);
+      kernel = op->getOperand(0);
+    } else {
+      input = op->getOperand(0);
+      kernel = op->getOperand(1);
+    }
+    Value output = op->getResult(0);
+
+    auto inputTy = input.getType().dyn_cast<RankedTensorType>();
+    auto kernelTy = kernel.getType().dyn_cast<RankedTensorType>();
+    auto outputTy = output.getType().dyn_cast<RankedTensorType>();
+
+    if (!inputTy || !kernelTy || !outputTy) return success();
+
+    SmallVector<int64_t> transposeAttr = {1, 0};
+
+    SmallVector<Value> newOperands;
+
+    // insert transpose before spgemm when sparse_weight_pos = 1
+    if (sparse_weight_pos == 1) {
+      auto inputShape = inputTy.getShape();
+      SmallVector<int64_t, 4> inputTranShape;
+      for (int64_t val : transposeAttr) {
+        inputTranShape.push_back(inputShape[val]);
+      }
+      auto newInputTy =
+          RankedTensorType::get(inputTranShape, inputTy.getElementType());
+
+      auto inputTranOp = InsertTranspose(op, input, transposeAttr, b);
+      inputTranOp->getResult(0).setType(newInputTy);
+      newOperands.push_back(inputTranOp->getResult(0));
+    } else {
+      newOperands.push_back(input);
+    }
+
+    newOperands.push_back(kernel);
+
+    // compute spgemm output shape
+    auto output_shape = outputTy.getShape();
```

Comment:
nit: I see multiple places using the same logic. maybe define a helper function to insert transpose op and calculate the new type.

Reply:
Done.